### ISSUE TITLE
Cloud Functions not deploying: Build fails

### ISSUE THREAD

**Issue by aravindvnair99** [firebase/firebase-functions#795]
### Version info

**node:** v10.18.1 (Issue is present irrespective of version)

**firebase-functions:** 3.6.1 (Issue is present irrespective of version)

**firebase-tools:** 8.12.1 (Issue is present irrespective of version)

**firebase-admin:** 9.2.0 (Issue is present irrespective of version)

### Steps to reproduce

I've been trying to deploy my existing project with just a few EJS changes. But it doesn't deploy even if I redeploy old code.

### Expected behavior

Should deploy

### Actual behavior

```
{"@type":"type.googleapis.com/google.cloud.audit.AuditLog","status":{"code":13,"message":"Build failed: Build error details not available. Please check the logs at https://console.cloud.google.com/logs/viewer?project=[Name removed]&advancedFilter=resource.type%3Dbuild%0Aresource.labels.build_id%3D[ID removed]%0AlogName%3Dprojects%2F[Name removed]%2Flogs%2Fcloudbuild"},"authenticationInfo":{"principalEmail":"[Email removed]"},"serviceName":"cloudfunctions.googleapis.com","methodName":"google.cloud.functions.v1.CloudFunctionsService.UpdateFunction","resourceName":"projects/[Name removed]/locations/us-central1/functions/app"} 
```

### Were you able to successfully deploy your functions?

No, I ended up with this:

```shell
⚠  functions[app(us-central1)]: Deployment error.
Build failed: Build error details not available. Please check the logs at https://console.cloud.google.com/logs/viewer?project=[Name removed]&advancedFilter=resource.type%3Dbuild%0Aresource.labels.build_id%3D[ID Removed]%0AlogName%3Dprojects%2F[Name removed]%2Flogs%2Fcloudbuild


Functions deploy had errors with the following functions:
	app


To try redeploying those functions, run:
    firebase deploy --only "functions:app"


To continue deploying other features (such as database), run:
    firebase deploy --except functions

Error: Functions did not deploy properly.

**Comment by aravindvnair99** [comment]
I fixed it by deleting us.artifacts.[Project ID].appspot.com

**Comment by richiricardo** [comment]
Hello, i am having this issue. Any idea on the solution?

**Comment by aravindvnair99** [comment]
> Hello, i am having this issue. Any idea on the solution?

@richiricardo The solution I mentioned works. Try it out.

**Comment by richiricardo** [comment]
How exactly do I do this ?

**Comment by aravindvnair99** [comment]
> How exactly do I do this ? 

@richiricardo You can go to Google Cloud Platform console, navigate to storage and then delete us.artifacts.[Project ID].appspot.com as I mentioned. Then redeploy. Should work fine.

**Comment by richiricardo** [comment]
Yo bro msg me priv ?

**Comment by aravindvnair99** [comment]
> Yo bro msg me priv ? 

@richiricardo Sorry mate, you just have to do what I said itself. That's the solution. Worked for others too! If you need additional help, you can reply here. For other issues, start a new issue tracker.

**Comment by richiricardo** [comment]
Is it a lot of lines of code ? Or just that one line to remove

**Comment by aravindvnair99** [comment]
> Is it a lot of lines of code ? Or just that one line to remove 

@richiricardo It's not a line of code. That's a storage bucket. You just need to navigate to the storage bucket I mentioned and delete the same in your project provided you have access for it. Else the concerned person has to do it for you.

**Comment by richiricardo** [comment]
There wasn’t any us.artifacts. Only gcf sources - staging -
projectid.appspot

**Comment by aravindvnair99** [comment]
> There wasn’t any us.artifacts. Only gcf sources - staging - projectid.appspot

@richiricardo [Project ID].appspot.com is the default bucket. As for GCF Sources, is `gcf-sources-[PROJECT_NUMBER]-[REGION]` the name of the bucket? If it's something else, you need to delete and recreate in this format. Also, you need one for every region you are deploying to.

**Comment by isamx** [comment]
Hello So am having the same problem as you guys 
I looked for the bucket that you minioned  but I couldn't find it
I only found 3:
 gcf-sources-537183344133-us-central1 
kodx-app.appspot.com 
staging.kodx-app.appspot.com

**Comment by richiricardo** [comment]
I redone the project and it worked.

**Comment by aravindvnair99** [comment]
@isamx Has the issue been resolved for you?

**Comment by Shekhar7860** [comment]
> I fixed it by deleting us.artifacts.[Project ID].appspot.com

where is this file

**Comment by aravindvnair99** [comment]
> > I fixed it by deleting us.artifacts.[Project ID].appspot.com
> 
> where is this file

@Shekhar7860 Please read the thread. I've already mentioned where to go and do what.

**Comment by veluvj** [comment]
> I fixed it by deleting us.artifacts.[Project ID].appspot.com

@aravindvnair99, The issue is still occurring after deleting  the bucket

**Comment by aravindvnair99** [comment]
> > I fixed it by deleting us.artifacts.[Project ID].appspot.com
> 
> @aravindvnair99, The issue is still occurring after deleting the bucket

@veluvj Which Node.js version are you using?

**Comment by veluvj** [comment]
> @veluvj Which Node.js version are you using?

@aravindvnair99   v10.23.3

**Comment by aravindvnair99** [comment]
> > @veluvj Which Node.js version are you using?
> 
> @aravindvnair99 v10.23.3

@veluvj Check my reply here: https://github.com/firebase/firebase-tools/issues/2699#issuecomment-787978223

**Comment by lllouiseF** [comment]
i just copy this on my one of my course in udemy: 

const functions = require('firebase-functions');
const cors = require('cors')({ origin: true });
const Busboy = require('busboy');
const os = require('os');
const path = require('path');
const fs = require('fs');
const uuid = require('uuid/v4');

const { Storage } = require('@google-cloud/storage');

const storage = new Storage({
  projectId: 'YOUR_FIREBASE_PROJECT_ID'
});

exports.storeImage = functions.https.onRequest((req, res) => {
  return cors(req, res, () => {
    if (req.method !== 'POST') {
      return res.status(500).json({ message: 'Not allowed.' });
    }
    const busboy = new Busboy({ headers: req.headers });
    let uploadData;
    let oldImagePath;

    busboy.on('file', (fieldname, file, filename, encoding, mimetype) => {
      const filePath = path.join(os.tmpdir(), filename);
      uploadData = { filePath: filePath, type: mimetype, name: filename };
      file.pipe(fs.createWriteStream(filePath));
    });

    busboy.on('field', (fieldname, value) => {
      oldImagePath = decodeURIComponent(value);
    });

    busboy.on('finish', () => {
      const id = uuid();
      let imagePath = 'images/' + id + '-' + uploadData.name;
      if (oldImagePath) {
        imagePath = oldImagePath;
      }

      console.log(uploadData.type);
      return storage
        .bucket('YOUR_FIREBASE_PROJECT_ID.appspot.com')
        .upload(uploadData.filePath, {
          uploadType: 'media',
          destination: imagePath,
          metadata: {
            metadata: {
              contentType: uploadData.type,
              firebaseStorageDownloadTokens: id
            }
          }
        })

        .then(() => {
          return res.status(201).json({
            imageUrl:
              'https://firebasestorage.googleapis.com/v0/b/' +
              storage.bucket('YOUR_FIREBASE_PROJECT_ID.appspot.com').name +
              '/o/' +
              encodeURIComponent(imagePath) +
              '?alt=media&token=' +
              id,
            imagePath: imagePath
          });
        })
        .catch(error => {
          console.log(error);
          return res.status(401).json({ error: 'Unauthorized!' });
        });
    });
    return busboy.end(req.rawBody);
  });
});


then i only changed the project name based on the professor..

**Comment by aravindvnair99** [comment]
> @aravindvnair99
> 
> i just copy this on my one of my course in udemy:
> // Code redacted
> then i only changed the project name based on the professor..

@lllouiseF As far as I have seen, the code of a project has never been the problem. As stated in my previous message, could you go to your Google Cloud Platform Console, navigate to Cloud Storage and list out the buckets you see?

**Comment by lllouiseF** [comment]
@aravindvnair99  i dont have any bucket yet. -.- what should i do?

**Comment by aravindvnair99** [comment]
> @aravindvnair99 i dont have any bucket yet. -.- what should i do?

@lllouiseF That doesn't sound right. You should be having buckets by default. Make sure you are in the right project. Could you attach a screenshot?

**Comment by lllouiseF** [comment]
@aravindvnair99  oh, i only have 2 buckets.

**Comment by aravindvnair99** [comment]
> @aravindvnair99 oh, i only have 2 buckets.

@lllouiseF Could you name the buckets or attach a screenshot?

Link if you need it: https://console.cloud.google.com/storage/ (Make sure you are in the right project)

**Comment by lllouiseF** [comment]
here 
![Screen Shot 2021-03-30 at 10 38 26 PM](https://user-images.githubusercontent.com/79818209/113007349-bad53400-91a8-11eb-83f9-cd3000e35d63.png)

**Comment by aravindvnair99** [comment]
> here
> ![Screen Shot 2021-03-30 at 10 38 26 PM](https://user-images.githubusercontent.com/79818209/113007349-bad53400-91a8-11eb-83f9-cd3000e35d63.png)

@lllouiseF Delete `us.artifacts.[Project ID].appspot.com` and try deploying again.

**Comment by lllouiseF** [comment]
@aravindvnair99  still the same :(

**Comment by aravindvnair99** [comment]
> @aravindvnair99 still the same :(

@lllouiseF Can you paste the content of your log here? (Run deploy with the debug flag)

**Comment by lllouiseF** [comment]
$ firebase deploy

Functions deploy had errors with the following functions:
        storeImage(us-central1)

To try redeploying those functions, run:
    firebase deploy --only "functions:storeImage"

To continue deploying other features (such as database), run:
    firebase deploy --except functions

Error: Functions did not deploy properly.

**Comment by aravindvnair99** [comment]
> $ firebase deploy
> 
> Functions deploy had errors with the following functions:
> storeImage(us-central1)
> 
> To try redeploying those functions, run:
> firebase deploy --only "functions:storeImage"
> 
> To continue deploying other features (such as database), run:
> firebase deploy --except functions
> 
> Error: Functions did not deploy properly.

@lllouiseF Ah... No, not like that. You missed the debug flag. Type `firebase --debug deploy`

**Comment by lllouiseF** [comment]
hey @aravindvnair99  i sent you a connection in linkedin. the message is too long -.-

**Comment by aravindvnair99** [comment]
> hey @aravindvnair99 i sent you a connection in linkedin. the message is too long -.-

@lllouiseF Ahh... You can copy-paste it into a text file and upload it here.

**Comment by lllouiseF** [comment]
i cant attach the file. can i just send it to your linkedin acc? accept my connection too.

**Comment by aravindvnair99** [comment]
> i cant attach the file. can i just send it to your linkedin acc? accept my connection too.

@lllouiseF Drag and drop into the GitHub comment box to attach the file. Alternatively, you can also copy-paste the file or click on the bottom bar which says `Attach files by dragging & dropping, selecting or pasting them.`

You can type `firebase deploy --only "functions:storeImage" --debug` to keep the log as short as possible. You only must upload the log for this one single command which should shed some light on why it's failing.

**Comment by lllouiseF** [comment]
[debug.txt](https://github.com/firebase/firebase-functions/files/6230346/debug.txt)
 @aravindvnair99

**Comment by lllouiseF** [comment]
@aravindvnair99  i'll get back to you tomorrow. It's very late here. Thank you for helping..

**Comment by aravindvnair99** [comment]
> [debug.txt](https://github.com/firebase/firebase-functions/files/6230346/debug.txt)
> @aravindvnair99 firebase --debug deploy

@lllouiseF Ah looks like you are deploying v12 and not v15 which is fine, but just wanted to point out. And the deploy seems fine. But your log says:

`error":{"code":3,"message":"Function failed on loading user code. This is likely due to a bug in the user code. Error message: Error: please examine your function logs to see the error cause: https://cloud.google.com/functions/docs/monitoring/logging#viewing_logs. Additional troubleshooting documentation can be found at https://cloud.google.com/functions/docs/troubleshooting#logging. Please visit https://cloud.google.com/functions/docs/troubleshooting for in-depth troubleshooting documentation."}`

Seems like your issue is different unlike the one others in this thread were discussing. I haven't checked your code yet, but could you give me a MCVE?

Refer this: https://stackoverflow.com/help/minimal-reproducible-example

> @aravindvnair99 i'll get back to you tomorrow. It's very late here. Thank you for helping..

Happy to help! Good night to you!

**Comment by lllouiseF** [comment]
hey @aravindvnair99  i debug it. Thank you again for your help!! Take care man!

**Comment by lllouiseF** [comment]
yow haha i have another problem lols.  Access to XMLHttpRequest at 'https://us-central1-ionic-angular-app-f437c.cloudfunctions.net/storeImage' from origin 'http://localhost:8100' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. @aravindvnair99  i already changed my storage to true so anyone could access

**Comment by aravindvnair99** [comment]
> yow haha i have another problem lols. Access to XMLHttpRequest at 'https://us-central1-ionic-angular-app-f437c.cloudfunctions.net/storeImage' from origin 'http://localhost:8100' has been blocked by CORS policy: No 'Access-Control-Allow-Origin' header is present on the requested resource. @aravindvnair99 i already changed my storage to true so anyone could access

@lllouiseF CORS isn't that. It stands for `Cross-Origin Resource Sharing`. Simply put, when you try to access data that is present on another origin, it should be a permitted origin. You haven't given me full details, but whichever resource you are trying to access should have the `HTTP Access-Control-Allow-Origin` header set.

Refer to the following for a better understanding:

- https://developer.mozilla.org/en-US/docs/Web/HTTP/CORS
- https://www.youtube.com/watch?v=4KHiSt0oLJ0

**Comment by lllouiseF** [comment]
![Screen Shot 2021-03-31 at 4 43 51 PM](https://user-images.githubusercontent.com/79818209/113116732-4e0d7880-9240-11eb-97e3-c78186b4d6a7.png)
@aravindvnair99

**Comment by aravindvnair99** [comment]
> Screen Shot 2021-03-31 at 4 43 51 PM
> @aravindvnair99

@lllouiseF https://stackoverflow.com/a/35553666/7023294

**Comment by rameshvishnoi90904** [comment]
> I fixed it by deleting us.artifacts.[Project ID].appspot.com

Thanks. This worked

**Comment by aravindvnair99** [comment]
> > I fixed it by deleting us.artifacts.[Project ID].appspot.com
> 
> Thanks. This worked

@rameshvishnoi90904 Glad, it was helpful! :)

I would recommend setting up automation for the deletion. Google seems to be doing it automatically now.

**Comment by simzhaolin** [comment]
> I fixed it by deleting us.artifacts.[Project ID].appspot.com

You kind sir. Thank you very much. Worked like a champ. Wish i discovered your solution earlier, wasted 4 hours on this issue. Thank you!

**Comment by aravindvnair99** [comment]
> > I fixed it by deleting us.artifacts.[Project ID].appspot.com
> 
> You kind sir. Thank you very much. Worked like a champ. Wish i discovered your solution earlier, wasted 4 hours on this issue. Thank you!

@simzhaolin Glad, it was helpful! :)

I would recommend setting up automation for the deletion. Google seems to be doing it automatically now.

**Comment by gagapg** [comment]
it does not work for me.
for my issue, it's because of `require('../service-account-file.json')`, so I need to remove all `require`
https://stackoverflow.com/a/62855195/4340131

**Comment by hungtranpg** [comment]
it does not work for me.
for my issue, it's because of `require('../service-account-file.json')`, so I need to remove all `require`
https://stackoverflow.com/a/62855195/4340131

**Comment by aravindvnair99** [comment]
> it does not work for me.
> for my issue, it's because of `require('../service-account-file.json')`, so I need to remove all `require`
> https://stackoverflow.com/a/62855195/4340131

@hungtranpg That's a different issue. This issue tracker is for a different issue. In your case, you don't need to remove all require. You just need to put the JSON in the correct folder. Or just use the environment variable. What are you trying to reference is already there in the environment as pointed out by the StackOverflow answer you linked.

**Comment by hungtranpg** [comment]
> You just need to put the JSON in the correct folder.

the JSON file, I place it outside of the "functions" folder so it might not be readable when deployed

>Or just use the environment variable

Yes, I'm using it. The link I linked in the StackOverflow refers to using it.

`import admin from "firebase-admin";`
`const firebaseAdminApp = admin.initializeApp();`

**Comment by hungtranpg** [comment]
@aravindvnair99 
> You just need to put the JSON in the correct folder.

the JSON file, I place it outside of the "functions" folder so it might not be readable when deployed

>Or just use the environment variable

Yes, I'm using it. The link I linked in the StackOverflow refers to using it.

`import admin from "firebase-admin";`
`const firebaseAdminApp = admin.initializeApp();`

**Comment by aravindvnair99** [comment]
> the JSON file, I place it outside of the "functions" folder so it might not be readable when deployed

@hungtranpg Yes, only that folder is packaged. If you keep it inside, it will work. But it's really not required as the 2nd option exists. Ideally you would want that file only in the development environment.

> Yes, I'm using it. The link I linked in the StackOverflow refers to using it.
> 
> `import admin from "firebase-admin";`
> `const firebaseAdminApp = admin.initializeApp();`

Yes, Google sets it automatically. You can use the same in your local system as well by setting the environment variable either at runtime or beforehand.

**Comment by clivi-telesalud** [comment]
Hi @aravindvnair99
I have the same problem. But each time I deploy my function, the us.artifacts.[Project ID].appspot.com paper again in my bucket.

**Comment by aravindvnair99** [comment]
> Hi @aravindvnair99
> I have the same problem. But each time I deploy my function, the us.artifacts.[Project ID].appspot.com appear again in my bucket.

@clivi-telesalud That's normal as it needs to build to deploy. You can read more on how Cloud Functions works to understand the process. You don't have to delete each time by the way. Just when it goes haywire is enough. You can also set the automatic emptying as I mentioned.

**Comment by i-am-mani** [comment]
@aravindvnair99 Can we delete `gcf-sources-<ID>-us-center1`? as i see it contains zipped source code for all the functions, but it also maintains previous version as well as those functions which i have deleted. So will it be okay to delete the entire `gcf-sources` folder and recreate everything by another round of deploy? or it is better to delete individual files(previous version and delete functions) which are not relevant anymore?

Thank you :)

**Comment by aravindvnair99** [comment]
> @aravindvnair99 Can we delete `gcf-sources-<ID>-us-center1`? as i see it contains zipped source code for all the functions, but it also maintains previous version as well as those functions which i have deleted. So will it be okay to delete the entire `gcf-sources` folder and recreate everything by another round of deploy? or it is better to delete individual files(previous version and delete functions) which are not relevant anymore?
> 
> Thank you :)

Hey @i-am-mani!

You just have to delete the contents of the bucket by selecting the main folder and then redeploy. If that doesn't help, then delete the bucket and it will automatically create it again for you on the next deployment. Things should be fine then.

**Comment by motiejunas** [comment]
@aravindvnair99 could you please reopen the issue?

Not the first time bumping to the same issue and today the manual workaround is not working (manually deleting buckets) anymore. In any case, does not sound like a good solution, let Firebase fix this properly...

**Comment by aravindvnair99** [comment]
@motiejunas Sure. I have actually created a ticket with Firebase for the same after facing it continuously now. Back in October 2020, when I mentioned the solution, it was a one time thing. But now I'm having to go delete bucket each time.

Here's my new GitHub issue: https://github.com/firebase/firebase-tools/issues/3359

Reopening this as well as per your suggestion and considering how many people are facing it.

**Comment by i-am-mani** [comment]
> > @aravindvnair99 Can we delete `gcf-sources-<ID>-us-center1`? as i see it contains zipped source code for all the functions, but it also maintains previous version as well as those functions which i have deleted. So will it be okay to delete the entire `gcf-sources` folder and recreate everything by another round of deploy? or it is better to delete individual files(previous version and delete functions) which are not relevant anymore?
> > Thank you :)
> 
> Hey @i-am-mani!
> 
> You just have to delete the contents of the bucket by selecting the main folder and then redeploy. If that doesn't help, then delete the bucket and it will automatically create it again for you on the next deployment. Things should be fine then.

Hey Thanks for your response. I meant can we delete GCF-Sources bucket **safely**? because on couple of places, its said that deleting the GCF bucket may cause problems. The problem with GCF bucket in my case  is that, it contains the functions that are no longer used and probably it contains history of each functions too.

Thanks again!

**Comment by aravindvnair99** [comment]
> Hey Thanks for your response. I meant can we delete GCF-Sources bucket **safely**? because on couple of places, its said that deleting the GCF bucket may cause problems. The problem with GCF bucket in my case is that, it contains the functions that are no longer used and probably it contains history of each functions too.
> 
> Thanks again!

@i-am-mani Emptying the artifacts is usually more than sufficient. If that doesn't work (like in my case), then delete the artifacts bucket. If that also doesn't work, empty the sources bucket. If that also doesn't work, then delete the sources bucket. It will automatically recreate as needed on deploy.

**Comment by taeold** [comment]
@aravindvnair99 Thank you so much for taking your time to help other users on deployment issues.

It looks like lots of users have found your advice of deleting the artifacts bucket useful. I'm unsure why this is the case, and I'll see what I can find by starting internal conversations.

Can someone share the error message they are seeing from their Cloud Build? Frankly, there isn't much I can pick up from reading error messages like this one:

```
Build failed: Build error details not available. Please check the logs at https://console.cloud.google.com/logs/viewer?project=[Name removed]&advancedFilter=resource.type%3Dbuild%0Aresource.labels.build_id%3D[ID Removed]%0AlogName%3Dprojects%2F[Name removed]%2Flogs%2Fcloudbuild
```

**Comment by aravindvnair99** [comment]
@taeold Happy to help! Although I don't face that issue anymore, I am facing an identical issue that I've mentioned at https://github.com/firebase/firebase-tools/issues/3359

My laptop's drive unfortunately ended up in a hard drive failure and won't boot. So I'm stuck with only my phone for the time being. Once I've got it fixed I'll see if I can get you my old logs. As far as I remember, it had something to do with the build not being able to find artifacts.

Meanwhile if others could share, that would be great!

**Comment by taeold** [comment]
@aravindvnair99 Does this look familiar? https://stackoverflow.com/questions/67338215/cloud-function-build-error-failed-to-get-os-from-config-file-for-image/67402286

I filed https://github.com/firebase/firebase-tools/issues/3404 intended to help with the overall situation with the artifacts being left behind in projects.

I'm leaning on closing this issue since it's grown to be a bucket of all kinds of deployment issue, and it's not clear what the requirements should be for us to close the issue. I'll leave the issue open for sometime to see if others can help us identify the root cause of this particular problem.

**Comment by motiejunas** [comment]
> @aravindvnair99 Does this look familiar? https://stackoverflow.com/questions/67338215/cloud-function-build-error-failed-to-get-os-from-config-file-for-image/67402286
> 
> I filed [firebase/firebase-tools#3404](https://github.com/firebase/firebase-tools/issues/3404) intended to help with the overall situation with the artifacts being left behind in projects.
> 
> I'm leaning on closing this issue since it's grown to be a bucket of all kinds of deployment issue, and it's not clear what the requirements should be for us to close the issue. I'll leave the issue open for sometime to see if others can help us identify the root cause of this particular problem.

**> it's not clear what the requirements should be for us to close the issue**

WHAT?? I cannot deploy my project for 2 weeks already. I see many other developers are experiencing the same. The requirement here is more than clear: "As a developer, I expect my project to be deployed when hitting "firebase deploy"."

People are paying money for that, cannot use the service properly and the response is like that, just wow.

The firebase/firebase-tools#3404 is yet another issue, but not the same. I want to deploy my project at least, growing storage issue is nothing compared.

**Comment by taeold** [comment]
@motiejunas Sorry if I didn't communicate well. I didn't mean to downplay your issues.

The concern that I was trying to raise is that this issue is too broad to offer any help. "My function deploy is failing" is so broad that I can't really offer any help without more details. There are many ways that a deployment may fail (e.g. GCP may be temporarily down, IAM permission issues, misconfigured user code to name a few), and I didn't like how this issue was becoming a bucket of all the things that could go wrong. I think it's better for issue to focus on a particular issue - that ways, others experiencing similar issue can find relevant information instead of trying out 10 different solutions suggested in this thread (see https://github.com/firebase/firebase-tools/issues/3359 for another example).

That being said, does https://stackoverflow.com/questions/67338215/cloud-function-build-error-failed-to-get-os-from-config-file-for-image/67402286 not help you?

tl;dr - Always delete container images from Cloud Container Registry directly. Don't delete cache image layers stored in Cloud Storage - it will break your subsequent function deploys.

**Comment by aravindvnair99** [comment]
# Reply for 1st part

> @aravindvnair99 Does this look familiar? https://stackoverflow.com/questions/67338215/cloud-function-build-error-failed-to-get-os-from-config-file-for-image/67402286

@taeold Yes, it does. That was the error. Tried checking earlier logs, but I was unable to find them. But as far as I remember, it is the same thing. But I checked for a few recent ones, and I noticed Cloud Functions says something like:

```JSON
{
  "protoPayload": {
    "@type": "type.googleapis.com/google.cloud.audit.AuditLog",
    "status": {
      "code": 3,
      "message": "Build failed: Build error details not available.Please check the logs at https://console.cloud.google.com/cloud-build/builds;region=us-central1/<redacted>?project=<redacted>. Please visit https://cloud.google.com/functions/docs/troubleshooting#build for in-depth troubleshooting documentation for build related errors."
    },
    "authenticationInfo": {
      "principalEmail": "<redacted>"
    },
    "serviceName": "cloudfunctions.googleapis.com",
    "methodName": "google.cloud.functions.v1.CloudFunctionsService.UpdateFunction",
    "resourceName": "projects/<redacted>/locations/us-central1/functions/app"
  },
  "insertId": "-1nctggc9uk",
  "resource": {
    "type": "cloud_function",
    "labels": {
      "region": "us-central1",
      "project_id": "<redacted>",
      "function_name": "app"
    }
  },
  "timestamp": "2021-06-14T04:43:07.310674Z",
  "severity": "ERROR",
  "logName": "projects/<redacted>/logs/cloudaudit.googleapis.com%2Factivity",
  "operation": {
    "id": "operations/<redacted>",
    "producer": "cloudfunctions.googleapis.com",
    "last": true
  },
  "receiveTimestamp": "2021-06-14T04:43:07.637300980Z"
}
```

Clicking on the link it gives, takes me to the build details where I can see that step 2 titled "analyzer" is failing :

```
2021-06-14 10:13:06.779 ISTStep #2 - "analyzer": [31;1mERROR: [0mfailed to initialize cache: failed to create image cache: accessing cache image "us.gcr.io/<redacted>/gcf/us-central1/<redacted>/cache:latest": failed to get OS from config file for image 'us.gcr.io/<redacted>/gcf/us-central1/<redacted>/cache:latest'
```

# Reply for 2nd part

> I'm leaning on closing this issue since it's grown to be a bucket of all kinds of deployment issue

The issue is the same for most people I believe (I have gone through multiple reports online and I have also helped many who approached me directly face to face / social media platforms). I first faced this issue and then opened this issue tracker on October 15th, 2020. I managed to figure it out on my own and closed it. And for me, the issue came up because I had set a life cycle for the cloud storage on my own. I did so as I ended up getting much higher-than-expected costs for Cloud Storage. Deleting the bucket fixed it for me as I mentioned here: https://github.com/firebase/firebase-functions/issues/795#issuecomment-710091616 and that was a one-time thing for all my projects sitting on Firebase. Also, this was only for Node.js 8 and 10. Node.js 12 did not seem to have this issue back then when I checked.

Later as time passed, I noticed Google was auto adding life cycles, especially to newer projects which was never the case before. So, I stopped manually setting it up. Everything was fine until now where suddenly everyone is facing issues. It is something that changed during that time that is causing the issue and that is when I opened https://github.com/firebase/firebase-tools/issues/3359 and is now occurring irrespective of the Node.js version. It has always been the Cloud Build that is failing due to the life cycle policies. Some recent change to the build process is what's causing all these issues. Deleting the bucket always fixes it for me. I have not tried deleting images yet. But as far as I can see, it's the life cycle that is creating issues for most people.

# Related issues:
- https://github.com/firebase/firebase-tools/issues/2699
- https://github.com/firebase/firebase-functions/issues/676
- https://issuetracker.google.com/issues/154260223
- https://github.com/firebase/firebase-tools/issues/3359
- https://github.com/firebase/firebase-tools/issues/2956

# Apart from the ones I listed, Stack Overflow is filled with questions related to:

- Cloud Function deployments failing due to Cloud Build fails
- Cloud Storage costs
- Purposes of storage buckets and their associated costs
- Cloud Storage lifecycle

# Concluding thoughts

But with that said, I feel properly documenting all necessary information in the right place and auto-deleting / cleaning up unnecessary stuff in the right manner after deployment should solve this issue for most if not all people because as far as I have seen, most people have life cycle policies set either manually or automatically from Google's end. I also feel duplicate issues can be closed so that everything can be tracked in a single place.

**Comment by taeold** [comment]
@aravindvnair99 Thanks for a great summary of what's happened so far. It helps me put things in larger perspective.

I think many folks have helped us understand what's missing in our documentations and in features. I still believe https://github.com/firebase/firebase-tools/issues/3404 will help the most, as it would help prevent users from having to worry about deleting container images (which, when done partially, could lead to messy state that breaks subsequent deployments).

One last puzzle:

> Later as time passed, I noticed Google was auto adding life cycles, especially to newer projects which was never the case before. So, I stopped manually setting it up.

This is a big surprise to me, and is the first time I'm hearing that something like this is happening without users knowledge. Can you share more detail? Around when did you started seeing this? On which buckets are you seeing this? Are you seeing it on every new project you create? (I tried creating a new project and deploying a function, and I didn't see a lifecycle policy setup on the artifacts buckets.)

Thanks a bunch for your help, here and elsewhere. I agree with you that we should beef up the documentation for advanced users like you prior to closing this issue w/o any real resolution.

**Comment by aravindvnair99** [comment]
> I still believe [firebase/firebase-tools#3404](https://github.com/firebase/firebase-tools/issues/3404) will help the most, as it would help prevent users from having to worry about deleting container images (which, when done partially, could lead to messy state that breaks subsequent deployments).

@taeold I too believe so! Hope to see it get merged soon.

> This is a big surprise to me, and is the first time I'm hearing that something like this is happening without users knowledge.

Oh, I thought it was a feature 😅 And I also found it useful as I no longer had to manually go and set, especially as some of my projects had 15GB+ worth of artifacts which weren't necessary at all.

> Can you share more detail? Around when did you started seeing this?

I no longer see it happening. I noticed it sometime after I raised this ticket. I don't remember which month exactly. It was last year though. I saw it for some old and as well as new projects. But after a while I stopped seeing it happen to new projects I was creating.

> On which buckets are you seeing this? Are you seeing it on every new project you create? (I tried creating a new project and deploying a function, and I didn't see a lifecycle policy setup on the artifacts buckets.)

This was for the artifacts bucket and was set to 3 days lifecycle. I always used to set to 1, but since 3 started coming automatically, I assumed that is the recommended value by Google. I no longer see it happening though. It was only during a brief period of last year I noticed the lifecycle being set for existing as well as new projects I was creating. For existing, I did confirm, with my team mates as well as I initially thought one of them did it. For new, I personally created.

Also, I just noticed one more thing... The created date for items in Container Registry say `1 Jan 1980`. That isn't right. You might want to look into that as well.

> Thanks a bunch for your help, here and elsewhere.

Happy to help! :)

**Comment by aravindvnair99** [comment]
@taeold Forgot to mention, I still do see the staging bucket getting a lifecycle of 15 days though. Apart from that, I no longer see lifecycle being set for any other bucket.

**Comment by taeold** [comment]
@aravindvnair99 Are any of your GCP projects (especially your staging one) under an organization policy? In my past troubleshooting sessions, I learned that org policy can automatically setup a lifecycle policy on storage buckets and leads to troubles w/ Google Cloud Functions.

**Comment by aravindvnair99** [comment]
> @aravindvnair99 Are any of your GCP projects (especially your staging one) under an organization policy? In my past troubleshooting sessions, I learned that org policy can automatically setup a lifecycle policy on storage buckets and leads to troubles w/ Google Cloud Functions.

@taeold Oh I see... Some of the projects are within an organization. Some of them aren't in my case. I've not set any such policies though. Just individual bucket lifecycle policies.

**Comment by taeold** [comment]
@aravindvnair99 IIUC, your individual bucket will inherit lifecycle policy setup by your org. I've seen a similar setup where anything tagged with "staging" will be cleaned up after X days, and was wondering if your staging project was under similar limitation. I haven't been able to find anyone internally who knows about automatic lifecycle on buckets yet. Will keep searching.

**Comment by aravindvnair99** [comment]
@taeold So far, for all the organisations I worked with, I was the one who setup all the cloud infrastructure. I never set-up such a policy. Is that policy set-up by default?

I could see it on both organisational and non organisational projects though.

**Comment by abin-m** [comment]
> > Hello, i am having this issue. Any idea on the solution?
> 
> @richiricardo The solution I mentioned works. Try it out.

i tried to delete that storage bucket, but it still shows the same error

**Comment by aravindvnair99** [comment]
> > > Hello, i am having this issue. Any idea on the solution?
> > 
> > 
> > @richiricardo The solution I mentioned works. Try it out.
> 
> i tried to delete that storage bucket, but it still shows the same error

@abin-m Delete the registry image and redeploy.

**Comment by abin-m** [comment]
> > > > Hello, i am having this issue. Any idea on the solution?
> > > 
> > > 
> > > @richiricardo The solution I mentioned works. Try it out.
> > 
> > 
> > i tried to delete that storage bucket, but it still shows the same error
> 
> @abin-m Delete the registry image and redeploy.

could you please elaborate
I can't find any registry images in my GCP.

**Comment by taeold** [comment]
@abin-m Container Registry images for your Firebase Functions should be visible on https://console.cloud.google.com/gcr. Please delete them (if you want to save storage cost) instead of deleting objects in your Google Cloud Storage to avoid breaking your deployment.

**Comment by taeold** [comment]
Also - consider upgrading your firebase cli to 9.14.0. We just shipped a feature where container images will be cleaned up after a successful function deploy!

https://github.com/firebase/firebase-tools/pull/3439

**Comment by inlined** [comment]
FYI: The staging bucket is a Google-created bucket that was originally used for App Engine uploads. It always has a lifecycle policy because it's only expected to be used by Google tools and those tools aren't ever supposed to be putting durable objects there (it is, after all, only a "staging" bucket). IIRC we stopped using the staging bucket for GCF a long time ago, so any issues involving the staging bucket should probably be closed with "upgrade your CLI".

Google Container Repository does not handle deleting the underlying GCS files cleanly. Adding a lifecycle hook will cause GCF and other tools to break. Any issues involving the artifacts bucket should probably be closed with "Don't mess with the artifacts bucket. Upgrade your CLI and we will safely purge caches for you"

Are there other outstanding issues that could be considered in-scope, or can this bug be closed? As @taeold commented, "not deploying" is too broad a scope for a single bug. Issues related to container registry failures fit here. Other specific issues/bugs can be new GitHub issues. If you can't narrow down your trouble deploying to any particular bug, [contacting support](https://firebase.google.com/support) will probably get you going faster

**Comment by abin-m** [comment]
Solved by downgrading Node js version to 12 and CLI also upgraded.

**Comment by aravindvnair99** [comment]
Closing as per discussions with @taeold at https://github.com/firebase/firebase-tools/issues/3359

**Comment by Garciconx** [comment]
Updating Node JS to version 16 work for me

`"engines": {
    "node": "16"
  },`

**Comment by bdrelling** [comment]
@Garciconx how did you get that to work? I couldn't deploy anything above Node 14 via Firebase Cloud Functions.

**Comment by taeold** [comment]
@bdrelling Go ahead and open a new issue if you are having problems. Logs would help!

**Comment by bdrelling** [comment]
I appreciate your help, but I'm not having any problems, so I'd rather not open an entirely new issue for a simple question in reply to a user on the same thread. Especially given that the [documentation](https://firebase.google.com/docs/functions/manage-functions#set_nodejs_version) does not show that Node 16 is supported yet, so it doesn't strike me as an issue at all. If someone confirms that Node 16 _can_ be used with Cloud Functions, then I'll happily open an issue.

I don't think there's value in stamping out discussion on issues if it could help potential users that continue to land on these PRs after the fact, especially while the issue is still open to comments. And if it's just a matter of noise, anyone can unsubscribe when they want to.

Apologies if I'm misunderstanding you.

**Comment by Garciconx** [comment]
> @Garciconx how did you get that to work? I couldn't deploy anything above Node 14 via Firebase Cloud Functions.

Sorry at the end I had to set the node version to 14, because 16th version leads to some errors with other libraries.

**Comment by bdrelling** [comment]
@taeold if I try to use Node 16 again, I'll be happy to open a new issue with any problems I run into. 👍🏻 

I totally get what you mean re: the team missing comments posted in an issue, and appreciate the redirection to open an issue (if I had one).

**Comment by DogeIsAwesome1** [comment]
i have the same error even deleted the archive could it maybe be my code or ?
``exports.addAlgoliaIndex = functions.firestore.document('order/document/serial{serial, productCode}').onCreate((change) => {
  const data = snapshot.data();
  const objectID = snapshot.id;
  return index.saveObject({ ...data, objectID });
});
exports.updateAlgoliaIndex = functions.firestore.document('order/document/serial{serial, productCode}').onUpdate((change) => {
  const newData = change.after.data();
  const objectID = change.after.id;
  return index.saveObject({ ...newData, objectID });``

**Comment by inlined** [comment]
@DogeIsAwesome1 Your document strings are invalid. {template} strings can only be full components of a slash and must be a valid identifier name. So something like `'order/document/serial/{serial}'`

**Comment by inlined** [comment]
Please file a support ticket at firebase.google.com/support

**Comment by nelsonlarocca** [comment]
hi !
any new about this ?
just moving to:

```
  "engines": {
    "node": "16"
  },
```
  
  in the package.json file throws 
  
  Error: package.json in functions directory has an engines field which is unsupported. Valid choices are: {"node": "10"}, {"node":"12"}, and {"node":"14"}.

Deploys to runtimes below Node.js 10 are now disabled in the Firebase CLI. Existing Node.js 8 functions will stop executing on 2021-03-15. Update existing functions to Node.js 10 or greater as soon as possible.

**Comment by inlined** [comment]
@nelsonlarocca Looks like you need a newer version of `firebase-tools`. Try running `npm install -g firebase-tools@latest` to get a version with support for Node 16.

**Comment by faustoct** [comment]
so what's the solution?

**Comment by faustoct** [comment]
> > so what's the solution?
> 
> @faustoct what's the issue you are facing?

it says: `✔  functions: functions folder uploaded successfully`

Then
`Functions deploy had errors with the following functions:`

````
i  functions: cleaning up build files...
Error: There was an error deploying functions:
````

all my functions show here... saying `Error Failed to update...`

**Comment by aravindvnair99** [comment]
> @aravindvnair99 any ideia

@faustoct you need to attach your log files for any of us to pin point your issue.

**Comment by faustoct** [comment]
> > @aravindvnair99 any ideia
> 
> @faustoct you need to attach your log files for any of us to pin point your issue.

@aravindvnair99 ok fixed

**Comment by inlined** [comment]
Fixed? I don't see a log file anywhere. I'll repeat again that this is a forum for reporting bugs, not troubleshooting general deploy problems. We have many more people employed to help you with the latter than the former. You'll find very qualified support at your disposal at [firebase.google.com/support](https://firebase.google.com/support).

**Comment by faustoct** [comment]
@inlined you are right. I am paid customer and forgot about the the support because I went through the support in Google Cloud dashboard instead Firebase Console.

Yes, I fixed!

**Comment by monkeyK1n9** [comment]
> > How exactly do I do this ?
> 
> @richiricardo You can go to Google Cloud Platform console, navigate to storage and then delete us.artifacts.[Project ID].appspot.com as I mentioned. Then redeploy. Should work fine.

Worked for me!

You need to open the project in the Google Cloud Platform console, then storage => bucket. You will see the [Project ID].appspot.com, delete it and open your code editor to run again `npm run deploy`
